okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.rules.TestRule
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+import java.net.InetAddress
+
+/** Apply this rule to tests that need an OkHttpClient instance. */
+class OkHttpClientTestRule : TestRule {
+  /**
+   * Returns an OkHttpClient for all tests to use as a starting point.
+   *
+   * The shared instance allows all tests to share a single connection pool, which prevents idle
+   * connections from consuming unnecessary resources while connections wait to be evicted.
+   *
+   * This client is also configured to be slightly more deterministic, returning a single IP
+   * address for all hosts, regardless of the actual number of IP addresses reported by DNS.
+   */
+  fun newClient(): OkHttpClient {
+    return OkHttpClient.Builder()
+        .connectionPool(connectionPool)
+        .dispatcher(dispatcher)
+        .dns(SINGLE_INET_ADDRESS_DNS) // Prevent unexpected fallback addresses.
+        .build()
+  }
+
+  fun newClientBuilder(): OkHttpClient.Builder = newClient().newBuilder()
+
+  fun ensureAllConnectionsReleased() {
+    connectionPool.evictAll()
+    assertThat(connectionPool.idleConnectionCount()).isEqualTo(0)
+  }
+
+  override fun apply(base: Statement, description: Description): Statement {
+    return object : Statement() {
+      override fun evaluate() {
+        try {
+          base.evaluate()
+        } finally {
+          ensureAllConnectionsReleased()
```

Comment:
This pretty much kills off concurrency in our test running, running this against a JVM singleton connectionPool.  Perhaps there is a way we can use internal access to the connection pool to count/cleanup only the relevant connections. Since connections are tied to Address, even making Dns a non singleton and tracking based on object identity would allow you to filter.

Reply:
Hmmmm. Yep, lemme to that.

Reply:
I changed it up to pool full OkHttpClient instances, which allows us concurrency and safety.